src/storage/migrationManager.ts

```diff
@@ -1,50 +1,111 @@
-    logger.info(
-      `Migrating global storage from ${globalStorageVersion} to ${CODEBASE_STORAGE_VERSION}`,
+    logger.debug(
+      `global storage version is incorrect: "${globalStorageVersion}", should be "${CODEBASE_STORAGE_VERSION}"`,
     );
-
-    // Add global migration logic here. Have fun with that.
-
-    // Stamp current global state
-    await manager.setGlobalState(DURABLE_STORAGE_VERSION_KEY, CODEBASE_STORAGE_VERSION);
+    await executeMigrations(globalStorageVersion, CODEBASE_STORAGE_VERSION, "global");
+    manager.setGlobalState(DURABLE_STORAGE_VERSION_KEY, CODEBASE_STORAGE_VERSION);
+  } else {
+    logger.debug(`global storage storage version is correct: "${globalStorageVersion}"`);
   }
+}
 
-  // also check workspace state for storage versioning. This may be first time running the extension
-  // in this workspace in quite some time.
+/** Check the storage version in **workspace state**, then migrate if necessary. */
+async function migrateWorkspaceState(manager: StorageManager): Promise<void> {
   const workspaceStorageVersion = await manager.getWorkspaceState<number>(
     DURABLE_STORAGE_VERSION_KEY,
   );
   if (workspaceStorageVersion !== CODEBASE_STORAGE_VERSION) {
-    logger.info(
-      `Migrating workspace storage from ${workspaceStorageVersion} to ${CODEBASE_STORAGE_VERSION}`,
+    logger.debug(
+      `workspace storage version is incorrect: "${workspaceStorageVersion}", should be "${CODEBASE_STORAGE_VERSION}"`,
+    );
+    await executeMigrations(workspaceStorageVersion, CODEBASE_STORAGE_VERSION, "workspace");
+    manager.setWorkspaceState(DURABLE_STORAGE_VERSION_KEY, CODEBASE_STORAGE_VERSION);
+  } else {
+    logger.debug(`workspace storage storage version is correct: v${workspaceStorageVersion}`);
+  }
+}
+
+/** Check the storage version in **secret storage**, then migrate if necessary. */
+async function migrateSecretStorage(manager: StorageManager): Promise<void> {
+  const secretStorageVersion: string | undefined = await manager.getSecret(
+    DURABLE_STORAGE_VERSION_KEY,
+  );
+  if (secretStorageVersion !== String(CODEBASE_STORAGE_VERSION)) {
+    logger.debug(
+      `secret storage version is incorrect: "${secretStorageVersion}", should be "${CODEBASE_STORAGE_VERSION}"`,
     );
+    await executeMigrations(Number(secretStorageVersion), CODEBASE_STORAGE_VERSION, "secret");
+    manager.setSecret(DURABLE_STORAGE_VERSION_KEY, String(CODEBASE_STORAGE_VERSION));
+  } else {
+    logger.debug(`secret storage version is correct: "${secretStorageVersion}"`);
+  }
+}
+
+/**
+ * Determine the migration versions/steps required to move between two storage versions, which may
+ * be multiple versions apart.
+ */
+function getMigrationVersions(
+  sourceVersion: number | undefined,
+  targetVersion: number,
+): {
+  versions: number[];
+  isUpgrade: boolean;
+} {
+  // either the source version or the lowest migration version (2)
+  const current: number = sourceVersion ?? Math.min(...migrationVersions);
```

Comment:
Why duplicate the number[] before passing into `Math.min()`? Not like it is going to mutate it.

Reply:
Not quite duplicating it; `Math.min` takes the equivalent of `*args`, so we can't pass the `number[]` in as-is and have to use the `...` spread operator

Reply:
Thanks for educating me!